### ISSUE TITLE
accuracy has always been zero

### ISSUE THREAD

**Issue by T1mu** [qqwweee/keras-yolo3#505]
When I pass the accuracy metric in the model.compile() function in **train.py**.
`model.compile(optimizer=Adam(lr=1e-4), loss={'yolo_loss': lambda y_true, y_pred:y_pred},metrics=['acc'])`

outputs:
![image](https://user-images.githubusercontent.com/51531133/60758441-3ee89f00-a049-11e9-9b70-97d6e14e23ed.png)

Plz help me

If acc can't be used.
how can I get the 'precision'?

**Comment by ZooZoo-tc** [comment]
I am having facing issue....

**Comment by sirius0503** [comment]
Accuracy, in this case will be (y_pred == y_true).sum()/ len(y_true) (, I'm representing this as numpy array formulation, but this will be tensors, so K.sum() and K.mean() would come into place), but what are the tensors y_true and y_pred, here? y_true represents the true class id, and y_pred is the loss value calculated from the final layer `yolo_loss` added to the conv-base, so accuracy is a meaningless metric here and since the loss value i.e, `y_pred` won't be equal to `y_true` ever you get 0 accuracy. You can have mAP as a metric here, which is in general used for object detection cases: 
[mAP for object Detection](https://medium.com/@jonathan_hui/map-mean-average-precision-for-object-detection-45c121a31173), plz read this article, if any question remains unanswered feel free to come by and ask.

**Comment by IrisDinge** [comment]
> Accuracy, in this case will be (y_pred == y_true).sum()/ len(y_true) (, I'm representing this as numpy array formulation, but this will be tensors, so K.sum() and K.mean() would come into place), but what are the tensors y_true and y_pred, here? y_true represents the true class id, and y_pred is the loss value calculated from the final layer `yolo_loss` added to the conv-base, so accuracy is a meaningless metric here and since the loss value i.e, `y_pred` won't be equal to `y_true` ever you get 0 accuracy. You can have mAP as a metric here, which is in general used for object detection cases:
> [mAP for object Detection](https://medium.com/@jonathan_hui/map-mean-average-precision-for-object-detection-45c121a31173), plz read this article, if any question remains unanswered feel free to come by and ask.

Thank for Kind reply, may I know which metrics can we use in this case to monitor the training process? if it's available, it's also better to callback each epoch. Thank U

**Comment by sirius0503** [comment]
@IrisDinge : If you were to define a metric function like below:

```
import keras.backend as K

def mean_pred(y_true, y_pred):
    return K.mean(y_pred)

model.compile(optimizer='rmsprop',
              loss='binary_crossentropy',
              metrics=['accuracy', mean_pred])
```

Taken from [Keras Custom Metrics](https://keras.io/metrics/)

It would be shown after each `epoch` end. Also, you can go by using keras `custom callbacks class`, where you can print the result using the `on epoch end` method. For object detection algorithms, [**mAP (mean average precision)**](https://medium.com/@jonathan_hui/map-mean-average-precision-for-object-detection-45c121a31173) the metric used for model comparison. 

Thus, you'll have to implement either a `keras custom metric` or a `keras custom callback`. In the case of a keras metric, you'll have to work with tensors and also keras backend, But, the best thing about this is that the computation inside will be part of the `computation graph` and won't be needed to run separately. With keras callbacks also you can add tensor computations to `computation graph` by using the [fetches argument](https://stackoverflow.com/questions/47079111/create-keras-callback-to-save-model-predictions-and-targets-for-each-batch-durin) of keras callbacks.

I personally implemented [IoU](https://www.pyimagesearch.com/2016/11/07/intersection-over-union-iou-for-object-detection/), by using keras callbacks, since I wasn't able to find a way to do it by using metrics. After, implementing IoU, **mAP** is an easy metric to get!
